proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,359 @@
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/698>`_.
+.. sectnum::
+.. contents::
+
+This proposal proposes extending ``-XQualifiedDo`` to literals, to enable more ergonomic and more powerful syntax than ``OverloadedStrings`` or ``OverloadedLists``.
+
+Motivation
+----------
+
+``OverloadedStrings`` and ``OverloadedLists`` are useful for writing normal string/list syntax that get desugared to non-built-in types. However, they do this via typeclasses, which inherits issues like type inference ambiguity. It's possible these issues are one reason why these extensions, despite being in GHC for a long time, are not defaults in GHC202X language editions. Furthermore, they could provide even more flexibility; one note in the original ``OverloadedLists`` `design <https://gitlab.haskell.org/ghc/ghc/-/wikis/overloaded-lists>`_ mentions the inability to support heterogeneous lists.
+
+These extensions allow extensibility of string and list literals, but there's currently no option to allow extending other literals, like numeric literals. Because the ``fromInteger`` function is a part of ``Num``, there's no way to overload numeric literals without providing a specification for all the ``Num`` operations. Even if we broke out ``fromInteger`` into a separate class, we still inherit the same typeclass issues as overloaded strings/lists.
+
+The initial motivation for this came from `String interpolation <https://github.com/ghc-proposals/ghc-proposals/pull/570>`_, where this syntax could provide a mechanism similar to Javascript's `tagged template literals <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates>`_. But it applies more generally to other literals as well.
+
+Semi-related to https://github.com/ghc-proposals/ghc-proposals/issues/438
+
+Proposed Change Specification
+-----------------------------
+
+High-level Overview
+~~~~~~~~~~~~~~~~~~~
+
+``-XQualifiedLiterals`` enables the following syntaxes for expressions
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``Foo.1``
+      - ``Foo.fromNatural 1``
+    * - ``Foo.(1)``
+      - ``Foo.fromNatural (1)``
+    * - ``Foo.(-1)``
+      - ``Foo.fromNegativeInteger (-1)``
+    * - ``Foo.(1.2)``
+      - ``Foo.fromRational (1.2)``
+    * - ``Foo."asdf"``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo."""asdf"""``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo.[x, y]``
+      - ``Foo.fromList (x `Foo.listCons` (y `Foo.listCons` Foo.listNil))``
+
+And the following syntaxes for patterns
+
+.. list-table::
+    :align: left
+
+    * - **New pattern syntax**
+      - **Desugared ViewPattern**
+    * - ``Foo.1``
+      - ``(Foo.toNatural -> 1)``
```

Comment:
We could do with some examples/motivation for these. In particular, I think `toNatural` etc. should be expected to return a `Maybe`, as with `listUncons`. 

Or an alternative for literals would be to desugar to `Foo.matchNatural 1 -> True`. That allows for cases where it is not convenient to convert the value back to a literal (which may or may not be worth the slight increase in complexity).

Reply:
Ah yes, I like matchNatural. Thanks!

Reply:
A problem with pattern matches, is how to mark them `COMPLETE`.

```haskell
listUncons -> Just ...
```

is partial match, but for heterogeneous or size-indexed lists the correct length vector is complete match.

Reply:
@phadej That's a good point, I forgot about `COMPLETE`. AFAICT the only mechanism we have to implement patterns without TH is ViewPatterns, which can't indicate completeness.

So even if we had perfect desugaring like below, we still wouldn't get completeness:
```hs
case hlist of
  -- sugared
  HList.[Identity (), Identity ()] -> ...
  -- desugared
  (magicMatch -> (Identity (), Identity ())) -> ...
```

I think `COMPLETE` is going to have to be out of scope, until a future proposal adding TH to this